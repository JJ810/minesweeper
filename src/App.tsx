import { useEffect, useState } from 'react';
import Button from '@mui/material/Button';
import { Box, MenuItem, Select, SelectChangeEvent, Typography } from '@mui/material';
import Board from './board/board';
import { sagaActions } from './store/sagaActions';
import { useAppDispatch, useAppSelector } from './store/hook';
import { mapSelector, messageSelector } from './board/boardSlice';
import { useAppStyles } from './common/appStyles';

function App() {
  const dispatch = useAppDispatch();
  const classes = useAppStyles();
  const map = useAppSelector(mapSelector);
  const message = useAppSelector(messageSelector);
  const [level, setLevel] = useState(1);

  useEffect(() => {
    dispatch({ type: sagaActions.INIT_GAME });
  }, []);

  const handleLevelChange = (e: SelectChangeEvent) => {
    setLevel(Number(e.target.value));
  };

  const handleStart = () => {
    dispatch({ type: sagaActions.START_GAME, payload: `new ${level}` });
  };

  return (
    <Box className={classes.root}>
      <Box className={classes.container}>
        <Box className={classes.controlsWrapper}>
          <Select
            className={classes.levelSelector}
            variant="standard"
            value={level.toString()}
            label="Level"
            onChange={handleLevelChange}>
            <MenuItem value={1}>Beginner</MenuItem>
            <MenuItem value={2}>Amateur</MenuItem>
            <MenuItem value={3}>Intermediate</MenuItem>
            <MenuItem value={4}>Expert</MenuItem>
          </Select>
          <Button variant="contained" data-testid="btn-start" onClick={handleStart}>
            {map.length ? 'Reset' : 'Start'}
          </Button>
        </Box>
        <Board map={map} />
        {message !== 'OK' && (
          <Typography variant="h5" mt={2}>
            {message}
          </Typography>
        )}
      </Box>
    </Box>
  );
}

export default App;
