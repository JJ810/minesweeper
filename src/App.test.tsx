import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { cleanup, render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';

import rootSaga from './store/sagas';
import App from './App';

const sagaMiddleware = createSagaMiddleware();
const mockStore = configureStore([sagaMiddleware]);

describe('App', () => {
  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  it('Should render correctly and match with snapshot', () => {
    const initialState = {
      board: {
        map: [],
        message: ''
      }
    };
    const store = mockStore(initialState);
    const tree = renderer
      .create(
        <Provider store={store}>
          <App />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render welcome message and start button', () => {
    const initialState = {
      board: {
        map: [],
        message: ''
      }
    };
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(wrapper.text().includes('Welcome! Press Start button to get started!')).toBe(true);
    expect(wrapper.text().includes('Start')).toBe(true);
  });

  it('Map render should change start button text to Reset', async () => {
    const initialState = {
      board: {
        map: [
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□'
        ],
        message: ''
      }
    };
    const store = mockStore(initialState);
    sagaMiddleware.run(rootSaga);

    const wrapper = render(
      <Provider store={store}>
        <App />
      </Provider>
    );

    expect(wrapper.getByText('Reset')).toBeTruthy();
  });
});
