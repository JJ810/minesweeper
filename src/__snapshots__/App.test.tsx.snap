// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should render correctly and match with snapshot 1`] = `
<div
  className="makeStyles-root-1 MuiBox-root css-0"
>
  <div
    className="makeStyles-container-2 MuiBox-root css-0"
  >
    <div
      className="makeStyles-controlsWrapper-3 MuiBox-root css-0"
    >
      <div
        className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary makeStyles-levelSelector-4 css-194ccxm-MuiInputBase-root-MuiInput-root-MuiSelect-root"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          className="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Beginner
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value="1"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <button
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
        data-testid="btn-start"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Start
      </button>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      Welcome! Press Start button to get started!
    </p>
    <h5
      className="MuiTypography-root MuiTypography-h5 css-1657z6w-MuiTypography-root"
    >
      
    </h5>
  </div>
</div>
`;
